### PR TITLE
feat: support context for resolve and resolveWeak

### LINKED ISSUE #4315: support context for resolveWeak
`require.resolveWeak("./" + fileA)`

- webpack-test/cases/chunks/weak-dependencies-context

### LINKED ISSUE #7830: [Bug]: Found `require.resolveWeak` on ESM module for `web` target build
### System Info

System:
    OS: macOS 14.6.1
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Memory: 32.71 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.14.0 - ~/.nvm/versions/node/v20.14.0/bin/node
    Yarn: 1.22.5 - /usr/local/bin/yarn
    npm: 10.7.0 - ~/.nvm/versions/node/v20.14.0/bin/npm
    pnpm: 9.9.0 - ~/.nvm/versions/node/v20.14.0/bin/pnpm
    Watchman: 2023.08.14.00 - /usr/local/bin/watchman
  Browsers:
    Chrome: 128.0.6613.120
    Safari: 17.6
  npmPackages:
    @rspack/cli: ^1.0.0 => 1.0.0 
    @rspack/core: ^1.0.0 => 1.0.0 

### Context
Some libraries like `@lodable/components` or `react-loadable` leverage this `require.resolveWeak` capability to implement their code-splitting abstractions. When generating ESM code that use them, we end up with `require.resolveWeak` inside ESM. This creates the runtime failure. 

### Issue

In a `production` Rspack build, a runtime failure was encountered in the browser due to a `require.resolveWeak` call in a ESM file. In the example, we've set `target: 'web'` and specified the target browsers in `swc-loader`:

`rspack-dist/479.js`
```js
    resolve (props) {
        if (true) {
            return require.resolveWeak(`./locale/${props.locale}/lite.js`);
        }
        return eval('require.resolve')(`./locale/${props.locale}/lite.js`);
    }
``` 

We are not encountering this issue in the Webpack build:

`wepack-dist/560.js`
```js
resolve (props) {
        if (true) {
            return /*require.resolve*/(__webpack_require__(268).resolve(`./${props.locale}/lite.js`));
        }
        return eval('require.resolve')(`./locale/${props.locale}/lite.js`);
    }
```

I've tried using Babel and encountered the same issue. Any thoughts or workarounds would be appreciated.
Seems similar to this issue in webpack: https://github.com/webpack/webpack/issues/18633.  Thanks in advance.

### Reproduce link

https://github.com/Pablinho/rspack-resolve-weak-issue

### Reproduce Steps

1.- `pnpm run build`

### PR DESCRIPTION
**Pull request by inottn**
## Summary
close #7830 close #4315 

## Checklist

- [x] Tests updated (or not required).
- [x] Documentation updated (or not required).

### CODE REVIEW

**Comment by LingyuCoder** [comment]
LGTM，cc @ahabhgk

File: crates/rspack_plugin_javascript/src/parser_plugin/common_js_imports_parse_plugin.rs [hunk 1]
```diff
@@ -47,23 +49,107 @@ fn create_commonjs_require_context_dependency(
...
+    weak: bool,
+  ) -> bool {
+    if param.is_string() {
       parser
         .dependencies
         .push(Box::new(RequireResolveDependency::new(
-          str.value.to_string(),
-          range,
+          param.string().to_string(),
+          param.range().into(),
```
Comment (ahabhgk) [inline crates/rspack_plugin_javascript/src/parser_plugin/common_js_imports_parse_plugin.rs:126]: I see.. I recommend moving -1 here, there are some mess about BasicEvaluatedExpression.range, we will refactor it, moving -1 here should easy for refactoring in the future

File: crates/rspack_plugin_javascript/src/parser_plugin/common_js_imports_parse_plugin.rs [hunk 2]
```diff
@@ -47,23 +49,107 @@ fn create_commonjs_require_context_dependency(
...
+    parser: &mut JavascriptParser,
+    param: &BasicEvaluatedExpression,
+    argument_expr: &Expr,
+    weak: bool,
+  ) {
+    let dep = create_require_resolve_context_dependency(
+      parser,
+      param,
+      argument_expr,
+      param.range().into(),
```
Comment (ahabhgk) [inline crates/rspack_plugin_javascript/src/parser_plugin/common_js_imports_parse_plugin.rs:148]: Same here

File: tests/webpack-test/cases/parsing/resolve-weak-context/test.filter.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
 
-module.exports = () => {return "https://github.com/web-infra-dev/rspack/issues/4315"}
+module.exports = () => {return true}
```
Comment (ahabhgk) [inline tests/webpack-test/cases/parsing/resolve-weak-context/test.filter.js:2]: You can delete this file

File: tests/webpack-test/cases/chunks/weak-dependencies-context/test.filter.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
 
-module.exports = () => {return "https://github.com/web-infra-dev/rspack/issues/4315"}
+module.exports = () => {return true}
```
Comment (ahabhgk) [inline tests/webpack-test/cases/chunks/weak-dependencies-context/test.filter.js:2]: You can delete this file

**Review by ahabhgk — COMMENTED** [review]
Generally looks good to me, some nitpicks need change

File: crates/rspack_plugin_javascript/src/parser_plugin/common_js_imports_parse_plugin.rs (hunk 1 above)
Comment (inottn) [reply crates/rspack_plugin_javascript/src/parser_plugin/common_js_imports_parse_plugin.rs]: Yes, it took me some time to realize that BasicEvaluatedExpression.range might not be entirely accurate.

File: crates/rspack_plugin_javascript/src/dependency/context/mod.rs [hunk 5]
```diff
@@ -72,3 +75,46 @@ fn context_dependency_template_as_require_call(
...
+    .get_module_graph()
+    .module_graph_module_by_dependency_id(id)
+    .is_none()
+  {
+    source.replace(range.start, range.end, &expr, None);
+    return;
+  }
+
+  for (content, start, end) in &dep.options().replaces {
+    source.replace(*start, *end - 1, content, None);
```
Comment (ahabhgk) [inline crates/rspack_plugin_javascript/src/dependency/context/mod.rs:110]: This should be also not needed, the range returned by create_context_dependency is already -1

Comment (inottn) [reply crates/rspack_plugin_javascript/src/dependency/context/mod.rs:110]: https://github.com/web-infra-dev/rspack/blob/146e0be6279376249e3502a88214f2d80d87a213/crates/rspack_plugin_javascript/src/dependency/context/mod.rs#L74

It doesn't seem so.

Comment (ahabhgk) [reply crates/rspack_plugin_javascript/src/dependency/context/mod.rs:110]: I see..

**Review by ahabhgk — APPROVED** [review]
Thank you!

**Pull request merged**